Лабораторная работа dev-j120 (Лысогора А. В.)

QAEn opened the pull request:
Николай Владимирович, здравствуйте! Посмотрите, пожалуйста, новые изменения в проекте.

QAEn committed:
first commit

QAEn committed:
Delete OrderAccountingSystemFX.iml

QAEn committed:
gitIgnore

QAEn committed:
Merge remote-tracking branch 'origin/development' into development

QAEn committed:
newCommit

QAEn committed:
lastChange

konivex commented on src/ru/avalon/j120/order_accounting_system/auxiliary_classes/Product.java:
```diff
@@ -33,7 +33,7 @@ public int getStockBalance() {
 
     public void setPrice(int setPrice) {
-        if(setPrice <= 0) System.err.println("There can't be a zero or negative price");
+        if(setPrice <= 0) System.err.println("There can't be a zero or negative price");// метод должен выбрасывать исключение
```
И где выброс исключения?

konivex commented on src/ru/avalon/j120/order_accounting_system/auxiliary_classes/WareHouse.java:
```diff
@@ -17,7 +17,8 @@ public WareHouse( ArrayList<Product> prod) {
         for(Product p : products)
             if(p.getArticleNumber().equals(product.getArticleNumber())){
-                JOptionPane.showMessageDialog(null, "The list almost contains the product with article " + product.getArticleNumber());
+                JOptionPane.showMessageDialog(null, "The list almost contains the product with article " + product.getArticleNumber(),
```
Должно выбрасываться исключение!
Класс WareHouse не должен взаимодействовать с пользователем, это не его зона ответственности.

konivex reviewed (changes requested):
Замечания

ListOfOrderItems.setHowManyOrdered:
    if (howManyOrdered > product.getStockBalance())
    Должно выбрасываться исключение.

OrderManager.adddOrderPosition(Product product, int howMany):
    if (listOfOrderItems.contains(product)
    Должно выбрасываться исключение.

Класс Order: не хватает set-методов для отдельных атрибутов (contactPhoneNumber, discountPercentage и т.п.); обратие внимание, что атрибуты должны меняться только для готовящихся заказов.

Класс Configuration: его реализация (наличие метода getProperties()) может позволить случайно изменить значение к.-л. свойства. Лучше определить метод для возвращения конкретного свойства по заданному имени.

Класс SerializationDeserializationOfOrder: 
- общее замечание: если при чтении/записи произойдёт сбой, то файл останется открытым;
- метод doDeserialization(): файла может не быть, если система запускается первый раз.

Отмечу, то замечания к классу SerializationDeserializationOfOrder не применимы к классу WriteReadProductToFromList, он реализован корректно.

Не ясно назначение класса OrderManager в вашей реализации.
Также не ясно, как будут обрабатываться позиции заказа в процессе его подготовки. Видимо, для этого и предназначен класс OrderManager, но тогда он его код надо исправить соответствующим образом. Обращу внимание, что позиции заказа можно менять (добавлять новые, удалять имеющиеся, менять количество заказываемых единиц в позиции) до тех пор, пока заказ находится в состоянии "Готовится" (об этом же замечание к классу Order).

QAEn committed:
lastChange

QAEn committed:
lastChange

QAEn committed:
lastChange

QAEn committed:
lastChange

LAV007 commented:
Николай Владимирович, здравствуйте! Можете, пожалуйста, посмотреть пакет ui, я в телеграмм-группе описал вопрос.

> Николай Владимирович, здравствуйте! Пытаюсь реализовать добавление позиции заказа в таблицу, через диалоговое окно,  но в таком исполнении метода .addRow() из класса MyDataDialog.java строчки в таблицу не добавляются.
> 
> Если задать добавление строки в таблицу напрямую при нажатии на кнопку Ok в диалоговом окне - строчка добавится, но лишь та, которую непосредственно задам (комментарии в классе MyAbstrackModalDialog). 
> 
> Подскажите пожалуйста, где может быть ошибка?

konivex commented:
В конструкторе `MyAbstractModalDialog(Frame owner, String title)` вызов конструктора предка д.б. таким: `super(owner, title, true);`. Я когда показывал этот класс на демо, то сделал ошибку, вызвав сначала конструктор предка с двумя параметрами, но ближе к концу демо исправился, сделав его именно таким как здесь.

LAV007 commented:
Спасибо!!))  как я так не внимательно?..  Пишет!)

QAEn committed:
lastChange in MyAbstractModalDialog changed parent constructor

LAV007 commented:
Николай Владимирович, я же правильно понимаю: в классе ListDemo.java необходимо обработать объект orders типа JTable, похожим образом, как это сделано с объектом ordersItems типа JTable, только внутри (при отображение на графическом интерфейсе) должны быть текстовые поля для ввода данных о заказчике.

QAEn committed:
Create two new button in ListDemo.java class ("Create customer..." and "Remove customer")

konivex commented:
Да, вы всё правильно понимаете, только данные должны обрабатываться не о "заказчике", а о заказе.

QAEn committed:
final changes

LAV007 commented:
Николай Владимирович, здравствуйте!
В последнем коммите по максимуму исправил ранее указанные недостатки. Не судите строго, сам понимаю, что работа совершенно сырая. Надеюсь такого минимального взаимодействия ui хватит для того, чтобы засчитать работу) и продолжить обучение на курсе.
Если быть честным, это, конечно, может быть мой недостаток, но я не совсем понимаю для чего учить (настолько плотно внедрять в модуль) swing, когда он уже давно не актуален. Он, может быть и является значительной составляющей стандартных пакетов, но точно не необходимым инструментом в современной разработке. Ввиду этого "непонимания", у меня просто напрочь отсутствовала какая-либо мотивация на то, чтобы выполнить это задание лучше, в сравнении с первым модулем, где я действительно приложил старания для выполнения лабораторных работ. Надеюсь, для будущих слушателей, учебный отдел разнообразит модули более свежими и востребованными на рынке библиотеками.  В противном случае, я просто не вижу смысла в старой программе. На мой взгляд, слушателей переподготовки необходимо готовить с прицелом на то, что после завершения обучения, выпускники будут востребованы на рынке, на не доучивать еще полгода самостоятельно новые библиотеки. Эти же мысли, я продублировал в отзыве за модуль, уж очень хочется, чтобы Политех был впереди планеты всей по качеству предоставляемых образовательных услуг!) Извините, за столь объемный комментарий. Спасибо!

konivex commented:
"ListOfOrderItems.setHowManyOrdered:
if (howManyOrdered > product.getStockBalance())
Должно выбрасываться исключение.": не исправлено.

"Класс Order: не хватает set-методов...": set-методы вы добавили, но не учли моё замечание "обратите внимание, что атрибуты должны меняться только для готовящихся заказов.". Исправьте.

SerializationDeserializationOfOrder.doDeserialization(): текущая логика работы метода такая: если файла нет, то создаём пустой ArrayList, а потом всё равно пытаемся читать файл. Думаете, такая логика верна? Так же замечу, что ссылку на созданный ArrayList вы сохраняете в локальную переменную, и теряете её тут же при выходе из блока "if(!file.exists())".

QAEn committed:
исправлено:
ListOfOrderItems.setHowManyOrdered:if (howManyOrdered > product.getStockBalance())
Должно выбрасываться исключение.";

"Класс Order: не хватает set-методов...": set-методы вы добавили, но не учли моё замечание "обратите внимание, что атрибуты должны меняться только для готовящихся заказов.";

SerializationDeserializationOfOrder.doDeserialization().

konivex commented:
SerializationDeserializationOfOrder.doDeserialization(): не в сторону исправили. Теперь когда у вас файла нет, то создаётся пустой файл, а потом вы из него читаете ArrayList, чего у вас не получится.

Исправьте последнее замечание.

QAEn committed:
исправлено:
class: SerializationDeserializationOfOrder

konivex commented:
Алексей Викторович,
Не исправлено. Превращение переменной в поле класса для метода SerializationDeserializationOfOrder.doDeserialization() не меняет ничего. Он по-прежнему не будет работать. Прочтите ещё раз моё замечание и подумайте, что нужно исправить в тексте метода, чтобы он стал работать как надо.

QAEn committed:
добавлено:
class: DataBase (j-130_LabWork#1)

QAEn committed:
method createNewTable() is PATTERN!

QAEn committed:
change method doDeserialization()

konivex reviewed (changes requested):
Для зачёта: добавьте в каждую таблицу две-три тестовые записи и сделайте так, чтобы приложение запускалось и показывало их в графическом интерфейсе.

Замечание: реализация класса DB (файл DB.java) будет приводить к большому количеству повисших соединений с базой данных. Исправьте.

QAEn committed:
Николай Владимирович, посмотрите пожалуйста изменения.

QAEn committed:
Исправлено подключение к базе данных. Рациональное использование ресурсов подключения.

QAEn committed:
!!!facepalm!!!
Изменения в методе addPositions()

LAV007 commented:
Николай Владимирович, видимо не только ЦП моего компа перегрелся... Исправил метод addPositions() из класса OrdersModel.java. Посмотрите, пожалуйста, когда будет время.

konivex reviewed (commented):
Выглядит хорошо.
Встретимся в тимс завтра (в вс) в три и посмотрим.

LAV007 commented:
Добро, понял. Буду завтра в 15.00 в Teams!

QAEn committed:
last change
